File: rust/glacier/src/widget/systray/menu.rs
use crate::{
    BlockOnTokio, color,
    menu::{self, entry::WithMenu},
    misc::icons,
    services::status_notifier as sni,
};

use menu::message::Message;
use snowcap_api::{
    signal::Signaler,
    widget::{self, Program, image, row, text},
};

pub type Menu = menu::Menu<Message>;
pub type Handle = menu::Handle<Message>;

enum Kind {
    Label,
    Radio(bool),
    Checkmark(bool),
}

struct Common {
    item_id: sni::item::UniqueId,
    node_id: i32,
    label: String,

    service: sni::StatusNotifier,
}

pub struct Entry {
    common: Common,
    kind: Kind,
}

pub struct SubMenu {
    common: Common,
    signaler: Signaler,
    config: menu::Config,
}

pub fn make_menu(
    sni_entries: &mut dyn Iterator<Item = (i32, &sni::menu::Properties)>,
    service: sni::StatusNotifier,
    item_id: &sni::item::UniqueId,
    config: &menu::Config,
) -> menu::Menu<Message> {
    let mut prev_is_sep = true;
    let mut menu = menu::Menu::new();

    for (node_id, props) in sni_entries {
        if !props.visible() {
            continue;
        } else if matches!(props.type_(), sni::menu::Type::Separator) {
            if !prev_is_sep {
                menu = menu.add_separator();
            }

            prev_is_sep = true;
            continue;
        }

        prev_is_sep = false;

        let mut entry = if props.children_display() == "submenu" {
            menu::Entry::menu(SubMenu::new(
                service.clone(),
                item_id,
                node_id,
                props,
                config.clone(),
            ))
        } else {
            menu::Entry::standard(Entry::new(service.clone(), item_id, node_id, props))
        };

        if !props.enabled() {
            entry = entry.disable();
        }

        menu = menu.add_entry(entry);
    }

    menu
}

impl Entry {
    pub fn new(
        service: sni::StatusNotifier,
        item_id: &sni::item::UniqueId,
        node_id: i32,
        props: &sni::menu::Properties,
    ) -> Self {
        let kind = match props.toggle_type() {
            sni::menu::ToggleType::None => Kind::Label,
            sni::menu::ToggleType::Radio => Kind::Radio(props.toggle_state().unwrap_or(false)),
            sni::menu::ToggleType::Checkmark => {
                Kind::Checkmark(props.toggle_state().unwrap_or(false))
            }
        };

        Self {
            common: Common::new(service, item_id, node_id, props),
            kind,
        }
    }
}

impl SubMenu {
    pub fn new(
        service: sni::StatusNotifier,
        item_id: &sni::item::UniqueId,
        node_id: i32,
        props: &sni::menu::Properties,
        config: menu::Config,
    ) -> Self {
        Self {
            common: Common::new(service, item_id, node_id, props),
            signaler: Signaler::new(),
            config,
        }
    }
}

impl Common {
    pub fn new(
        service: sni::StatusNotifier,
        item_id: &sni::item::UniqueId,
        node_id: i32,
        props: &sni::menu::Properties,
    ) -> Self {
        Self {
            item_id: item_id.to_owned(),
            node_id,
            label: props.label().to_owned(),

            service,
        }
    }

    pub fn view_label(&self) -> snowcap_api::widget::WidgetDef<Message> {
        let label = text::Text::new(&self.label).width(widget::Length::Fill);

        label.into()
    }

    pub fn click(&self) {
        self.service.click_menu(&self.item_id, self.node_id)
    }

    pub fn hover(&self) {
        self.service.hover_menu(&self.item_id, self.node_id)
    }
}

impl Program for Entry {
    type Message = Message;

    fn view(&self) -> Option<snowcap_api::widget::WidgetDef<Self::Message>> {
        let mut children = vec![self.common.view_label()];

        let toggle_icon = match self.kind {
            Kind::Checkmark(state) => Some(icons::checkmark::select(state)),
            Kind::Radio(state) => Some(icons::radio::select(state)),
            _ => None,
        };

        if let Some(toggle_icon) = toggle_icon {
            let handle = toggle_icon.to_image_handle(Some(color::from_hex("#ffffff")));

            let icon = image::Image::new(handle)
                .content_fit(image::ContentFit::ScaleDown)
                .height(widget::Length::Fixed(16.))
                .width(widget::Length::Fixed(16.));

            children.push(icon.into());
        }

        let row = row::Row::new_with_children(children);
        Some(row.into())
    }

    fn update(&mut self, msg: Self::Message) {
        use menu::entry::Message as EMessage;

        match msg {
            Message::Entry(EMessage::Hover) => {
                self.common.hover();
            }
            Message::Entry(EMessage::Submit) => {
                self.common.click();
            }
            _ => {}
        }
    }
}

impl Program for SubMenu {
    type Message = Message;

    fn view(&self) -> Option<widget::WidgetDef<Self::Message>> {
        Some(self.common.view_label())
    }

    fn signaler(&self) -> Option<Signaler> {
        Some(self.signaler.clone())
    }

    fn update(&mut self, msg: Self::Message) {
        use menu::entry::Message as EMessage;

        if let Message::Entry(EMessage::Hover) = msg {
            self.common.hover()
        }
    }
}

impl WithMenu for SubMenu {
    fn open_menu(&self) -> Option<menu::Menu<Self::Message>> {
        let service = &self.common.service;
        let item_id = &self.common.item_id;
        let node_id = self.common.node_id;

        let (tx, rx) = tokio::sync::oneshot::channel();

        service.open_menu(item_id, node_id, {
            let config = self.config.clone();
            let service = service.clone();
            let item_id = item_id.to_owned();

            move |iter| {
                let menu = make_menu(iter, service, &item_id, &config);

                if tx.send(menu).is_err() {
                    tracing::error!("Failed to send menu");
                }
            }
        });

        rx.block_on_tokio().ok()
    }
}

File: lua/glacier/widget/systray/menu.lua
local Widget = require("snowcap.widget")

local _base = require("snowcap.widget.base")

local _menu = require("glacier.menu")
local color = require("glacier.misc.color")
local icons = require("glacier.misc.icons")

---@class glacier.widget.systray.menu
local menu = {}

---------------------
-- Type definition --
---------------------

---@class glacier.widget.systray.menu.Common
---@field item_id string
---@field node_id integer
---@field label string
---
---@field service glacier.services.StatusNotifier

---@class glacier.widget.systray.menu.Kind
---@field label? {}
---@field radio? boolean
---@field checkmark? boolean

---@class glacier.widget.systray.menu.Entry: glacier.widget.systray.menu.Common, snowcap.widget.Program
---@field kind glacier.widget.systray.menu.Kind
local Entry = setmetatable({}, { __index = _base.Base })

---@class glacier.widget.systray.menu.SubMenu: glacier.widget.systray.menu.Common, glacier.menu.entry.WithMenu
---@field config glacier.menu.PopupConfig
local SubMenu = setmetatable({}, { __index = _base.Base })

----------------------
-- Modules function --
----------------------

---@param sni_entries glacier.services.status_notifier.MenuNode[]
---@param service glacier.services.StatusNotifier
---@param item_id string
---@param config glacier.menu.PopupConfig
---@return glacier.menu.Menu
function menu.make_menu(sni_entries, service, item_id, config)
    local prev_is_sep = true
    local entries = {}

    for _, node in ipairs(sni_entries) do
        if not node.properties.visible then
            goto continue
        elseif node.properties.type == "separator" then
            if not prev_is_sep then
                table.insert(entries, _menu.entry.separator())
            end

            prev_is_sep = true
            goto continue
        end

        prev_is_sep = false

        if node.properties["children-display"] == "submenu" then
            local submenu = SubMenu.new(service, item_id, node.id, node.properties, config)
            local entry = _menu.entry.menu(submenu, {
                disabled = not node.properties.enabled,
            })
            table.insert(entries, entry)
        else
            local _entry = Entry.new(service, item_id, node.id, node.properties)
            local entry = _menu.entry.standard(_entry, {
                disabled = not node.properties.enabled,
            })
            table.insert(entries, entry)
        end

        ::continue::
    end

    return _menu({
        entries = entries,
    })
end

----------------
-- Entry Impl --
----------------

---------------------------------
-- impl snowcap.widget.Program --
---------------------------------

---Creates a widget definition for display by Snowcap.
---
---A widget may return nil to notify its parent program that it has
---nothing to display. It's up to the parent to decide whether to display a
---placeholder or to remove the widget from the tree.
---
---@return snowcap.widget.WidgetDef?
function Entry:view()
    local children = {
        Widget.text({
            text = self.label,
            width = Widget.length.Fill,
        }),
    }

    local toggle_icon
    if self.kind.checkmark ~= nil then
        toggle_icon = self.kind.checkmark ~= nil and icons.checkbox.select(self.kind.checkmark)
    elseif self.kind.radio ~= nil then
        toggle_icon = self.kind.radio ~= nil and icons.radio.select(self.kind.radio)
    end

    if toggle_icon then
        local handle = toggle_icon:to_image_handle(color.from_hex("#FFFFFF"))
        local icon = Widget.Image({
            handle = handle,
            content_fit = Widget.image.content_fit.SCALE_DOWN,
            height = Widget.length.Fixed(16.),
            width = Widget.length.Fixed(16.),
        })

        table.insert(children, icon)
    end

    return Widget.row({
        children = children,
    })
end

---Updates this widget program with the received message.
---@param msg any|glacier.menu.entry.Message
function Entry:update(msg)
    if msg.tag ~= _menu.entry.MESSAGE_TAG then
        return
    end

    local event = msg.event
    if not event then
        return
    end

    if event.hover then
        self.service:hover_menu(self.item_id, self.node_id)
    elseif event.submit then
        self.service:click_menu(self.item_id, self.node_id)
    end
end

---Called when a surface has been created with this program.
---
---A surface handle is provided to allow the program to manupulate
---the surface. This handle should be passed to any child programs
---to allow them to use it as well.
---
---@param _ snowcap.widget.SurfaceEvent
function Entry:event(_) end

--------------
-- Lifetime --
--------------

---@param service glacier.services.StatusNotifier
---@param item_id string
---@param node_id integer
---@param properties glacier.protocols.status_notifier.layout.Properties
---
---@return glacier.widget.systray.menu.Entry
function Entry.new(service, item_id, node_id, properties)
    local base = _base.Base.new()
    local ret = setmetatable(base, { __index = Entry }) --[[@as glacier.widget.systray.menu.Entry]]

    ret.service = service
    ret.item_id = item_id
    ret.node_id = node_id
    ret.label = properties.label

    if properties["toggle-type"] == "radio" then
        ret.kind = {
            radio = properties["toggle-state"] == 1,
        }
    elseif properties["toggle-type"] == "checkmark" then
        ret.kind = {
            checkmark = properties["toggle-state"] == 1,
        }
    else
        ret.kind = {
            label = {},
        }
    end

    return ret
end

------------------
-- SubMenu Impl --
------------------

---------------------------------
-- impl snowcap.widget.Program --
---------------------------------

---Creates a widget definition for display by Snowcap.
---
---A widget may return nil to notify its parent program that it has
---nothing to display. It's up to the parent to decide whether to display a
---placeholder or to remove the widget from the tree.
---
---@return snowcap.widget.WidgetDef?
function SubMenu:view()
    return Widget.text({
        text = self.label,
    })
end

---Updates this widget program with the received message.
---@param msg any|glacier.menu.entry.Message
function SubMenu:update(msg)
    if msg.tag ~= _menu.entry.MESSAGE_TAG then
        return
    end

    local event = msg.event
    if not event then
        return
    end

    if event.hover then
        self.service:hover_menu(self.item_id, self.node_id)
    end
end

---Called when a surface has been created with this program.
---
---A surface handle is provided to allow the program to manupulate
---the surface. This handle should be passed to any child programs
---to allow them to use it as well.
---
---@param _ snowcap.widget.SurfaceEvent
function SubMenu:event(_) end

--- Called by glacier.menu.Entry to retrieve the menu.
function SubMenu:open_menu()
    local service = self.service
    local item_id = self.item_id
    local node_id = self.node_id

    local nodes = service:open_menu(item_id, node_id)
    if nodes then
        return menu.make_menu(nodes, service, item_id, self.config)
    end
end

--------------
-- Lifetime --
--------------

---@param service glacier.services.StatusNotifier
---@param item_id string
---@param node_id integer
---@param properties glacier.protocols.status_notifier.layout.Properties
---@param config glacier.menu.PopupConfig
function SubMenu.new(service, item_id, node_id, properties, config)
    local base = _base.Base.new()
    local ret = setmetatable(base, { __index = SubMenu }) --[[@as glacier.widget.systray.menu.SubMenu]]

    ret.service = service
    ret.item_id = item_id
    ret.node_id = node_id
    ret.label = properties.label
    ret.config = config

    return ret
end

-----------
-- Other --
-----------

return menu
